File: sheep/sheepFileReader.cpp
#include <sstream>
#include <iostream>
#include <fstream>
#include "sheepFileReader.h"
#include "sheepCodeGenerator.h"
#include "sheepCodeBuffer.h"
#include "sheepTypes.h"
#include "sheepException.h"
#include "sheepImportTable.h"
#include "sheepMachine.h"
#include "sheepCaseInsensitiveStringCompare.h"


SheepFileReader::SheepFileReader(const std::string& filename)
{
	std::stringstream output;

	std::ifstream file(filename.c_str(), std::ios_base::binary);
	if (!file)
	{
		throw SheepException("Unable to open input file", SHEEP_ERR_FILE_NOT_FOUND);
	}

	unsigned int fileSize = getFileSize(file);
	//byte* data = new byte[fileSize];
	byte* data = SHEEP_NEW_ARRAY(byte, fileSize);
	file.read((char*)data, fileSize);
	file.close();

	read(data, fileSize);
	delete[] data;
}

SheepFileReader::SheepFileReader(std::ifstream& file)
{
	unsigned int fileSize = getFileSize(file);
	//byte* data = new byte[fileSize];
	byte* data = SHEEP_NEW_ARRAY(byte, fileSize);
	file.read((char*)data, fileSize);
	file.close();

	read(data, fileSize);
	delete[] data;
}

SheepFileReader::SheepFileReader(const byte* data, int length)
{
	read(data, length);
}


#define READ4(d,i) { memcpy(d, &data[i], 4); i+=4; }
#define READ2(d,i) { memcpy(d, &data[i], 2); i+=2; }
#define READ1(d,i) { memcpy(d, &data[i], 1); i+=1; }
#define READN(d,i,n) { memcpy(d, &data[i], n); i+=n; }

void SheepFileReader::read(const byte* data, unsigned int length)
{
	Sheep::SheepLanguageVersion languageVersion = Sheep::SheepLanguageVersion::V100;

	// read the header
	SheepHeader header;
	unsigned int offset = 0;
	READ4(&header.Magic1, offset);
	READ4(&header.Magic2, offset);
	READ4(&header.Magic3, offset);

	if (header.Magic1 != SheepHeader::Magic1Value || header.Magic2 != SheepHeader::Magic2Value || header.Magic3 != 0)
	{
		throw SheepException("Input file is not a valid sheep file", SHEEP_ERR_INVALID_FILE_FORMAT);
	}

	READ4(&header.ExtraOffset, offset);
	READ4(&header.DataOffset, offset);
	READ4(&header.DataSize, offset);
	READ4(&header.DataCount, offset);

	//header.OffsetArray = new unsigned int[header.DataCount];
	header.OffsetArray = SHEEP_NEW_ARRAY(unsigned int, header.DataCount);

	for (unsigned int i = 0; i < header.DataCount; i++)
		READ4(&header.OffsetArray[i], offset);

	if (header.ExtraOffset != header.DataOffset)
	{
		// read the "extra" header info
		SheepHeaderExtra extra;
		READ2(&extra.VersionMajor, offset);
		READ2(&extra.VersionMinor, offset);
		READ4(&extra.Checksum, offset);
		READ2(&extra.Year, offset);
		READ2(&extra.Month, offset);
		READ2(&extra.DayOfWeek, offset);
		READ2(&extra.Day, offset);
		READ2(&extra.Hour, offset);
		READ2(&extra.Minute, offset);
		READ2(&extra.Second, offset);
		READ2(&extra.Milliseconds, offset);
		
		// we don't care what the copyright info says, so just read until we find the end
		char c;
		do 
		{
			READ1(&c, offset);
		} while (c != 0);

		// update the version
		if (extra.VersionMajor == 1 && extra.VersionMinor == 0)
			languageVersion = Sheep::SheepLanguageVersion::V100;
		else if (extra.VersionMajor == 2 && extra.VersionMinor == 0)
			languageVersion = Sheep::SheepLanguageVersion::V200;
		else
		{
			// unknown version! we don't support it!
			throw SheepException("Input file is not a supported version of Sheep", SHEEP_ERR_INVALID_FILE_FORMAT);
		}
	}

	m_intermediateOutput = SHEEP_NEW IntermediateOutput(languageVersion);

	SectionHeader importHeader, constantsHeader, functionsHeader, variablesHeader, codeHeader;

	for (unsigned int i = 0; i < header.DataCount; i++)
	{
		if (header.OffsetArray[i] >= length)
		{
			delete[] header.OffsetArray;
			throw SheepException("Input file is not a valid sheep file", SHEEP_ERR_INVALID_FILE_FORMAT);
		}

		// seek
		offset = header.DataOffset + header.OffsetArray[i];
	
		char label[13] = {0};
		READN(label, offset, 12);
		if (std::string(label) == "SysImports")
		{
			int bytesRead = 0;
			importHeader = readSectionHeader(data + offset, label, &bytesRead);
			offset += bytesRead;
			unsigned int currentOffset = offset;

			for (unsigned int j = 0; j < importHeader.DataCount; j++)
			{
				SheepImport import;
				
				// seek
				offset = currentOffset + importHeader.OffsetArray[j];

				short lengthOfName;
				byte numReturns, numParameters;

				READ2(&lengthOfName, offset);
				import.Name = readString(data + offset, lengthOfName, &bytesRead);
				offset += bytesRead + 1;
				READ1(&numReturns, offset);
				READ1(&numParameters, offset);
				
				import.ReturnType = (SheepSymbolType)numReturns;

				for (byte k = 0; k < numParameters; k++)
				{
					byte parameterType;
					READ1(&parameterType, offset);

					if (parameterType == (byte)SheepSymbolType::Int ||
						parameterType == (byte)SheepSymbolType::Float ||
						parameterType == (byte)SheepSymbolType::String)
						import.Parameters.push_back((SheepSymbolType)parameterType);
				}
					
				m_intermediateOutput->Imports.push_back(import);
			}
		}
		else if (std::string(label) == "StringConsts")
		{
			int bytesRead = 0;
			constantsHeader = readSectionHeader(data + offset, "StringConsts", &bytesRead);
			offset += bytesRead;
	
			unsigned int currentOffset = offset;

			for (unsigned int j = 0; j < constantsHeader.DataCount; j++)
			{
				// seek
				offset = currentOffset + constantsHeader.OffsetArray[j];

				SheepStringConstant constant;
				constant.Offset = constantsHeader.OffsetArray[j];
				constant.Value = readString(data + offset, &bytesRead);
				offset += bytesRead;

				m_intermediateOutput->Constants.push_back(constant);
			}
		}
		else if (std::string(label) == "Variables")
		{				
			int bytesRead = 0;
			variablesHeader = readSectionHeader(data + offset, "Variables", &bytesRead);
			offset += bytesRead;
			
			unsigned int currentOffset = offset;
			
			for (unsigned int j = 0; j < variablesHeader.DataCount; j++)
			{
				// seek
				offset = currentOffset + variablesHeader.OffsetArray[j];
				
				unsigned short len;
				READ2(&len, offset);

				SheepSymbol symbol;
				symbol.Name = readString(data + offset, len+1, &bytesRead);
				offset += bytesRead;

				unsigned int type, value;
				READ4(&type, offset);
				READ4(&value, offset);

				if (type == (unsigned int)SheepSymbolType::Int)
				{
					symbol.Type = SheepSymbolType::Int;
					symbol.InitialIntValue = offset;
				}
				else if (type == (unsigned int)SheepSymbolType::Float)
				{
					symbol.Type = SheepSymbolType::Float;
					symbol.InitialFloatValue = offset;
				}
				else if (type == (unsigned int)SheepSymbolType::String)
				{
					symbol.Type = SheepSymbolType::String;
					symbol.InitialStringValue = offset;
				}
				
				m_intermediateOutput->Symbols.push_back(symbol);
			}
		}
		else if (std::string(label) == "Functions")
		{
			int bytesRead = 0;
			functionsHeader = readSectionHeader(data + offset, "Functions", &bytesRead);
			offset += bytesRead;
			unsigned int currentOffset = offset;
			
			for (unsigned int j = 0; j < functionsHeader.DataCount; j++)
			{
				// seek
				offset = currentOffset + functionsHeader.OffsetArray[j];
			
				SheepFunction func(NULL);

				unsigned short len;
				READ2(&len, offset);

				func.Name = readString(data + offset, len+1, &bytesRead);
				offset += bytesRead;

				char returnType, numParameters;
				READ1(&returnType, offset);
				READ1(&numParameters, offset);

				func.ReturnType = (SheepSymbolType)returnType;
				for (int k = 0; k < numParameters; k++)
				{
					char paramType;
					READ1(&paramType, offset);

					SheepSymbol param;
					param.Type = (SheepSymbolType)paramType;
					func.Parameters.push_back(param);
				}
				
				READ4(&func.CodeOffset, offset);

				func.ParentCode = m_intermediateOutput;
				m_intermediateOutput->Functions.push_back(func);
			}
		}
		else if (std::string(label) == "Code")
		{
			int bytesRead = 0;
			codeHeader = readSectionHeader(data + offset, "Code", &bytesRead);
			offset += bytesRead;
			unsigned int currentOffset = offset;

			if (codeHeader.DataCount > 1)
				throw SheepException("Extra code sections found", SHEEP_ERR_INVALID_FILE_FORMAT);

			// seek
			offset = currentOffset + codeHeader.OffsetArray[0];
			currentOffset = offset;
			
			std::vector<SheepFunction>& functions = m_intermediateOutput->Functions;
			for (unsigned int j = 0; j < functions.size(); j++)
			{	
				// seek
				offset = currentOffset + functions[j].CodeOffset;
				unsigned int size = 0;
				if (j == functions.size() - 1)
					size = codeHeader.DataSize - functions[j].CodeOffset;
				else
					size = functions[j+1].CodeOffset - functions[j].CodeOffset;

				functions[j].Code = SHEEP_NEW SheepCodeBuffer(size);
				functions[j].Code->Write((const char*)data + offset, size); 
			}
		}
		else
		{
			throw SheepException("Unrecognized data section", SHEEP_ERR_INVALID_FILE_FORMAT);
			
		}
	}

	// clean up
	delete[] header.OffsetArray;
	if (importHeader.OffsetArray) delete[] importHeader.OffsetArray;
	if (constantsHeader.OffsetArray) delete[] constantsHeader.OffsetArray;
	if (variablesHeader.OffsetArray) delete[] variablesHeader.OffsetArray;
	if (functionsHeader.OffsetArray) delete[] functionsHeader.OffsetArray;
	if (codeHeader.OffsetArray) delete[] codeHeader.OffsetArray;
}

void SheepFileReader::WireImportCallbacks(SheepMachine* vm)
{
}

SectionHeader SheepFileReader::readSectionHeader(const byte* data, const std::string& name, int* bytesRead)
{
	SectionHeader header;
	
	int offset = 0;
	strncpy(header.Label, name.c_str(), 12);
	READ4(&header.ExtraOffset, offset);
	READ4(&header.DataOffset, offset);
	READ4(&header.DataSize, offset);
	READ4(&header.DataCount, offset);

	//header.OffsetArray = new unsigned int[header.DataCount];
	header.OffsetArray = SHEEP_NEW_ARRAY(unsigned int, header.DataCount);

	for (unsigned int i = 0; i < header.DataCount; i++)
		READ4(&header.OffsetArray[i], offset);

	if (bytesRead) *bytesRead = offset;
	return header;
}

std::string SheepFileReader::readString(const byte* data, int* bytesRead)
{
	char c;
	std::string str;
	
	int offset = 0;

	c = (char)data[offset++];
	while(c != 0)
	{
		str.push_back(c);
		c = (char)data[offset++];
	}

	if (bytesRead) *bytesRead = offset;
	return str;
}

std::string SheepFileReader::readString(const byte* data, unsigned int length, int* bytesRead)
{
	char c;
	std::string str;
	
	int offset = 0;

	for (unsigned int i = 0; i < length; i++)
	{
		c = (char)data[offset++];
		if (c == 0) break;
			
		str.push_back(c);
	}

	if (bytesRead) *bytesRead = offset;
	return str;
}

unsigned int SheepFileReader::getFileSize(std::ifstream& file)
{
	std::ifstream::pos_type pos = file.tellg();
	file.seekg(0, std::ios_base::end);

	std::ifstream::pos_type end = file.tellg();
	file.seekg(pos, std::ios_base::beg);

	return end;
}

File: sheep/extra/SheepVMDotNet/SheepVMDotNet/SheepFileReader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SheepVMDotNet
{
    public class SheepFileReader
    {
        private struct SheepHeader
        {
            public uint Magic1;
            public uint Magic2;
            public uint Unknown;
            public uint ExtraOffset;
            public uint DataOffset;
            public uint DataSize;
            public uint DataCount;

            public uint[] OffsetArray;

            public const uint SheepHeaderSize = 28;
            public const uint Magic1Value = 0x53334b47;
            public const uint Magic2Value = 0x70656568;
        }

        private struct SectionHeader
        {
            public string Label;
            public uint ExtraOffset;
            public uint DataOffset;
            public uint DataSize;
            public uint Datacount;

            public uint[] OffsetArray;

            public const uint SectionHeaderSize = 28;
        }

        private IntermediateOutput _intermediateOutput;

        public SheepFileReader(string filename)
        {
            using (System.IO.Stream stream = new System.IO.FileStream(filename, System.IO.FileMode.Open))
            {
                read(stream);
            }
        }

        public IntermediateOutput Output
        {
            get { return _intermediateOutput; }
        }

        private void read(System.IO.Stream stream)
        {
            _intermediateOutput = new IntermediateOutput();

            System.IO.BinaryReader reader = new System.IO.BinaryReader(stream);

            SheepHeader header;
            header.Magic1 = reader.ReadUInt32();
            header.Magic2 = reader.ReadUInt32();

            if (header.Magic1 != SheepHeader.Magic1Value ||
                header.Magic2 != SheepHeader.Magic2Value)
                throw new Exception("Input file is not a valid sheep file");

            header.Unknown = reader.ReadUInt32();
            header.ExtraOffset = reader.ReadUInt32();
            header.DataOffset = reader.ReadUInt32();
            header.DataSize = reader.ReadUInt32();
            header.DataCount = reader.ReadUInt32();

            header.OffsetArray = new uint[header.DataCount];

            for (uint i = 0; i < header.DataCount; i++)
                header.OffsetArray[i] = reader.ReadUInt32();

            for (uint i = 0; i < header.DataCount; i++)
            {
                if (header.OffsetArray[i] >= stream.Length)
                    throw new Exception("Input file is not a valid Sheep file");

                stream.Seek(header.DataOffset + header.OffsetArray[i], System.IO.SeekOrigin.Begin);

                SectionHeader sectionHeader = readSectionHeader(reader);
                long currentOffset = stream.Position;

                if (sectionHeader.Label == "SysImports")
                {
                    for (uint j = 0; j < sectionHeader.Datacount; j++)
                    {
                        SheepImport import;

                        stream.Seek(currentOffset + sectionHeader.OffsetArray[j], System.IO.SeekOrigin.Begin);

                        short lengthOfName = reader.ReadInt16();
                        import.Name = readString(reader, lengthOfName);
                        import.Callback = null;

                        // skip padding
                        reader.ReadByte();

                        byte numReturns = reader.ReadByte();
                        byte numParameters = reader.ReadByte();

                        import.ReturnType = (SheepSymbolType)numReturns;
                        import.Parameters = new SheepSymbolType[numParameters];

                        for (byte k = 0; k < numParameters; k++)
                        {
                            SheepSymbolType paramterType = (SheepSymbolType)reader.ReadByte();

                            if (paramterType == SheepSymbolType.Int ||
                                paramterType == SheepSymbolType.Float ||
                                paramterType == SheepSymbolType.String)
                                import.Parameters[k] = paramterType;
                        }
                        
                        _intermediateOutput.Imports.Add(import);
                    }
                }
                else if (sectionHeader.Label == "StringConsts")
                {
                    for (uint j = 0; j < sectionHeader.Datacount; j++)
                    {
                        stream.Seek(currentOffset + sectionHeader.OffsetArray[j], System.IO.SeekOrigin.Begin);

                        SheepStringConstant constant;
                        constant.Offset = sectionHeader.OffsetArray[j];
                        constant.Value = readString(reader);

                        _intermediateOutput.Constants.Add(constant);
                    }
                }
                else if (sectionHeader.Label == "Variables")
                {
                    for (uint j = 0; j < sectionHeader.Datacount; j++)
                    {
                        stream.Seek(currentOffset + sectionHeader.OffsetArray[j], System.IO.SeekOrigin.Begin);

                        SheepSymbol symbol = new SheepSymbol();
                        short len = reader.ReadInt16();
                        symbol.Name = readString(reader, len);

                        // skip padding
                        reader.ReadByte();

                        symbol.Type = (SheepSymbolType)reader.ReadUInt32();

                        if (symbol.Type == SheepSymbolType.Int)
                            symbol.InitialIntValue = reader.ReadInt32();
                        else if (symbol.Type == SheepSymbolType.Float)
                            symbol.InitialFloatValue = reader.ReadSingle();
                        else if (symbol.Type == SheepSymbolType.String)
                            symbol.InitialStringValue = reader.ReadInt32();
                        else
                            throw new Exception("???");

                        _intermediateOutput.Symbols.Add(symbol);
                    }
                }
                else if (sectionHeader.Label == "Functions")
                {
                    for (uint j = 0; j < sectionHeader.Datacount; j++)
                    {
                        SheepFunction func = new SheepFunction();
                        short len = reader.ReadInt16();
                        func.Name = readString(reader, len);

                        reader.ReadByte();
                        reader.ReadByte();
                        func.CodeOffset = reader.ReadUInt32();

                        _intermediateOutput.Functions.Add(func);
                    }
                }
                else if (sectionHeader.Label == "Code")
                {
                    if (sectionHeader.Datacount > 1)
                        throw new Exception("Extra code sections found");

                    for (int j = 0; j < _intermediateOutput.Functions.Count; j++)
                    {
                        stream.Seek(currentOffset + _intermediateOutput.Functions[j].CodeOffset, System.IO.SeekOrigin.Begin);

                        uint size;
                        if (j == _intermediateOutput.Functions.Count - 1)
                            size = sectionHeader.DataSize - _intermediateOutput.Functions[j].CodeOffset;
                        else
                            size = _intermediateOutput.Functions[j + 1].CodeOffset - _intermediateOutput.Functions[j].CodeOffset;

                        SheepFunction f = _intermediateOutput.Functions[j];
                        f.Code = reader.ReadBytes((int)size);
                        _intermediateOutput.Functions[j] = f;
                    }
                }
            }
        }

        private static SectionHeader readSectionHeader(System.IO.BinaryReader reader)
        {
            SectionHeader header;

            byte[] label = reader.ReadBytes(12);
            header.Label = System.Text.Encoding.ASCII.GetString(label).Trim((char)0);
            header.ExtraOffset = reader.ReadUInt32();
            header.DataOffset = reader.ReadUInt32();
            header.DataSize = reader.ReadUInt32();
            header.Datacount = reader.ReadUInt32();

            header.OffsetArray = new uint[header.Datacount];

            for (uint i = 0; i < header.Datacount; i++)
            {
                header.OffsetArray[i] = reader.ReadUInt32();
            }

            return header;
        }

        private static string readString(System.IO.BinaryReader reader, short length)
        {
            byte[] bytes = reader.ReadBytes(length);
            return System.Text.Encoding.ASCII.GetString(bytes);
        }

        private static string readString(System.IO.BinaryReader reader)
        {
            StringBuilder b = new StringBuilder();

            long start = reader.BaseStream.Position;
            long end = start;

            byte c = reader.ReadByte();
            while (c != 0)
            {
                end++;
                c = reader.ReadByte();
            }

            reader.BaseStream.Seek(start, System.IO.SeekOrigin.Begin);
            byte[] buffer = reader.ReadBytes((int)(end - start));

            return System.Text.Encoding.ASCII.GetString(buffer);

            return reader.ReadString();
        }
    }
}
